experiments/experiment_prototype.py

```diff
@@ -1,385 +1,1246 @@
+        """
+        Add a slice to the experiment.
+        :param exp_slice: a slice (dictionary of slice_keys) to add to the experiment.
+        :param interfacing_dict: dictionary of type {slice_id : INTERFACING , ... } that defines how
+         this slice interacts with all the other slices currently in the experiment.
+        :return: the slice_id of the new slice that was just added.
+        """
+        if not isinstance(exp_slice, dict):
+            # TODO error log
+            return
+        exp_slice['slice_id'] = self.new_slice_id
+        # each added slice has a unique id, even if previous slices have been deleted.
+        exp_slice['cpid'] = self.cpid
+        new_exp_slice = self.setup_slice(exp_slice)
+        # check for any errors after defaults have been filled.
+        print('Requested Add {}'.format(exp_slice))
+        print('Adding (with Defaults) {}'.format(new_exp_slice))
+        self.__slice_dict[new_exp_slice['slice_id']] = new_exp_slice
+        for ind in self.slice_ids:
+            if ind == new_exp_slice['slice_id']:
+                continue
+            try:
+                self._interface[(ind, new_exp_slice['slice_id'])] = interfacing_dict[ind]
+                # update to add interfacing. new slice_id will be greater than all others so
+                # we can add with ind first and maintain interfacing list rule of key1 < key2.
+            except TypeError or IndexError:
+                self._interface[(ind, new_exp_slice['slice_id'])] = None
+                print('Interfacing not Fully Updated - Will Cause Errors so Please Update.')
+                # TODO return a warning if interfacing dictionary not updated at this time.
+
+        return new_exp_slice['slice_id']
+
+    def del_slice(self, remove_slice_id):
+        """
+        Remove a slice from the experiment.
+        :param remove_slice_id: the id of the slice you'd like to remove.
+        :return: boolean True if successful
+        """
+        if isinstance(remove_slice_id, int) and remove_slice_id in self.slice_ids:
+            del(self.slice_dict[remove_slice_id])
+            for key1, key2 in self._interface.keys():
+                if key1 == remove_slice_id or key2 == remove_slice_id:
+                    del self._interface[(key1, key2)]
+            return True
+        else:
+            return False
+            # TODO log that it cannot be removed
+            # errmsg = 'Cannot remove slice id {} : it does not exist'.format(remove_slice_id)
+            # raise ExperimentException(errmsg)
 
-    def build_Scans(self): # REVIEW #40 Function name does not adhere to PEP. Should not have caps.
-        """ REVIEW #7 Just a note to update this documentation away from controlprogram
-        Will run after a controlprogram instance is set up and
-        modified - build iterable objects for radar_control
+    def edit_slice(self, edit_slice_id, param1, value1, param2=None, value2=None, param3=None,
```

Comment:
I think it goes along with the design to give unique slice IDs to the slices and keep track of your parameters (new params = new slice = new slice ID). Plus, I think it's best to have to go through all the error checking again as in add_slice. I think it's just a good idea to prevent errors and also could be useful in writing your experiment. In your experiment you know that slice 2 was certain params, slice 3 was the same except 1 param, etc. and you could use that information inside your experiment as well.